src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs

```diff
@@ -326,20 +326,232 @@ private void MakeAndAssignEnvironments()
                         {
                             closure.CapturesThis = true;
                         }
                     }
+
+                    closuresCapturingScopeVariables.Add(scope, capturingClosures);
                 });
+
+                _closuresCapturingScopeVariables = closuresCapturingScopeVariables;
+            }
+
+            /// <summary>
+            /// Must be called only after <see cref="MakeAndAssignEnvironments"/> and <see cref="ComputeLambdaScopesAndFrameCaptures"/>
+            /// 
+            /// In order to reduce allocations, merge environments into a parent environment when it is safe to do so.
+            /// This must be done whilst preserving semantics.
+            /// 
+            /// We also have to make sure not to extend the life of any variable.
+            /// This means that we can only merge an environment into its parent if exactly the same closures directly or indirectly reference both environments.
+            /// </summary>
+            private void MergeEnvironments()
+            {
+                // for now we don't analyze jumps to check if it is safe to optimize, and instead return.
+                if (containsJumps(ScopeTree.BoundNode.Syntax))
+                    return;
+
+                // if a closure captures a scope, which captures its parent, then the closure also captures the parents scope.
+                // we update _closuresCapturingScopeVariables to reflect this.
+                foreach (var (scope, capturingClosures) in _closuresCapturingScopeVariables)
+                {
+                    if (scope.DeclaredEnvironments.Count == 0)
+                        continue;
+                    // Right now we only create one environment per scope
+                    Debug.Assert(scope.DeclaredEnvironments.Count == 1);
+                    var env = scope.DeclaredEnvironments[0];
+
+                    if (env.CapturesParent)
+                    {
+                        var currentScope = scope.Parent;
+                        while (true)
+                        {
+                            if (currentScope == null)
+                            {
+                                throw ExceptionUtilities.Unreachable;
+                            }
+
+                            if (currentScope.DeclaredEnvironments.Count == 0 || currentScope.DeclaredEnvironments[0].IsStruct)
+                            {
+                                currentScope = currentScope.Parent;
+                                continue;
+                            }
+
+                            _closuresCapturingScopeVariables[currentScope].AddAll(capturingClosures);
+
+                            // Right now we only create one environment per scope
+                            Debug.Assert(scope.DeclaredEnvironments.Count == 1);
+                            if (currentScope.DeclaredEnvironments[0].CapturesParent)
```

Comment:
You should be able to write this as a `while (env.CapturesParent)` instead of `while (true)` and set env to the new parent here.